ktxspec.adoc

```diff
@@ -1079,14 +1083,21 @@ as described in Section ??? of <<KDF13>>. Images in this format are
 a strict RGB-only ASTC HDR subset that must be transcoded to BC6H
 or decoded to raw 16-bit floating-point values before being uploaded
 to and sampled by a GPU that does not support ASTC HDR. UASTC HDR
-images can be supercompressed with Zstandard (`supercompressionScheme`
-= 2). If supercompression is used, the DFD's `bytesPlane[0-7]`
-must be set to 0 as described in the next subsection.
+images can be supercompressed with any scheme except BasisLZ
+(`supercompressionScheme` = 1). If supercompression is used, the
+DFD's `bytesPlane[0-7]` must be set to 0 as described in the next
+subsection.
 +
 This color model has a single channel with id
 `KHR_DF_CHANNEL_UASTC_HDR_RGB` (= 0) for which the
 `KHR_DF_SAMPLE_DATATYPE_FLOAT` flag (= 0x80) must be set.
 
+[NOTE]
+====
+Although the Vulkan format names for ASTC HDR use `SFLOAT`, ASTC HDR and
+therefore UASTC HDR values cannot be < 0.
```

Comment:
As confirmed in https://gitlab.khronos.org/vulkan/vulkan/-/issues/3983, ASTC HDR values can be negative. The note should address only UASTC HDR.

Reply:
Thanks @lexaknyazev. I'll wait for @fluppeteer to report back from checking Imagination's and Arm's implementations before I update this WIP.

Reply:
How should we handle sampleLower and the `SIGNED` qualifier in the DFD sample? I think we should use the same values as full ASTC (sampleLower = 0xBF800000U — IEEE 754 floating-point representation for -1.0f — and `SIGNED` set) because

1. Validators won't need to special case for vkFormat `VK_FORMAT_ASTC_*_SFLOAT` and colorModel 167.
2. The value of the `SIGNED` qualifier in this case makes no difference to how code should determine the magnitude of the of the sample. The only downside I see is potentially wasted checks of the sign bit.

Reply:
Realistically, implementations would just look at the `vkFormat` and `colorModel` to understand that it's UASTC HDR. That said, I don't see the reasons above compelling enough:
1. The mere presence of two DFD color models for `ASTC_4x4_SFLOAT` creates a special case anyway. On top of this, validators ideally should check that block data is compatible with UASTC HDR restrictions. Right now, the transcoder exits immediately on any error found. An alternative approach may be to define and output "error color" or "undefined" BC6H blocks for incompatible input blocks.
2. The purpose of DFD is to be an accurate and self-contained description of pixel data. This is certainly more applicable to raw formats and one could even say that most DFD fields are redundant for compressed color models and therefore should be ignored but I'd prefer standard's self-consistency in this case. Since UASTC HDR is its own color model, its DFD should accurately reflect UASTC HDR features.

Reply:
So you want the `SIGNED` qualifier unset and `sampleLower` = 0 — IEEE 754 floating-point representation for 0?

Reply:
Reporting back, Arm confirm that they preserve whatever's in the void extent block even if it's negative (thank you for the test case!), and say that NVIDIA do too, although I assume they must have tested on GL4.6 since I don't think there's a hardware decoder for NVIDIA. We believe IMG's hardware also preserves it, but we're confirming that our simulator matches the hardware before we're definitive. I'm waiting to hear back from Samsung. Thanks for raising all this - it should have been in the Vulkan CTS, so I'll make sure it is, but it doesn't look like a spec change will be happening, and the Vulkan ASTC enums will remain SFLOAT.

I think I agree that UASTC should have SAMPLE_LOWER set to 0, though. I imagine the signed bit should still be set, by symmetry with ASTC, but I don't feel strongly either way. Obviously an unsigned float value has a different representation, but we're not accessing it in raw form here regardless.

Intending to have some stuff merged for review by the end of the weekend. Apologies again for the hold up(s).

Reply:
After looking again at the DFD definitions, I think they're quite restrictive and we may be forced to follow ASTC HDR semantics.

The `FLOAT` qualifier must be set because UASTC HDR represents floating-point values as per ASTC HDR and BC6H specs.

The qualifier's description says:
   > Unless a genuine unsigned format is intended, `KHR_DF_SAMPLE_DATATYPE_SIGNED` (bit **_S_**) should also be set.

   The standard `float16` encoding is inherently signed - it has one sign bit, five exponent bits, and ten mantissa bits. A "genuinely" unsigned compatible format would have to be defined as `float15` with one padding bit. I don't think that such complication is justified unless directly defined in the DFD spec.

The expressions for applying `sampleLower` and `sampleUpper` values explicitly depend on the signedness. It would be clearly wrong to use `samplerLower = 0` with the `SIGNED` qualifier set. Following this logic, the DFD example for BC6H unsigned is incorrect, by the way.

---

So there are only two options:

- Set the `SIGNED` qualifier to false, set `sampleLower` to `0.0`, and therefore diverge from the description of the `FLOAT` qualifier.
- Set the `SIGNED` qualifier to true, set `sampleLower` to `-1.0`, and therefore imply that negative values could hypothetically be supported.

Reply:
KDFS does have a definition of unsigned 16-bit floats, as used in ufloat16 by Vulkan - section 10.1. I think it would complicate life for anyone trying to feed ASTC void-extent blocks into tooling, though (the one place a float16 is actually a float16 in ASTC, it's not...), so I'm not sure that helps.

My gut says to set signed to true but sampleLower to 0.0, but my gut is rarely sensible and I've not given this as much thought as others have...

Reply:
> KDFS does have a definition of unsigned 16-bit floats

That section merely defines how to parse sign, exponent, and mantissa bits from a "raw" 16-bit value. It does not define an unsigned 16-bit float type, which would likely have 11 mantissa bits (or one padding bit).

> to set signed to true but sampleLower to 0.0

This would make the sampled values incorrect, see the equations in Section 5.17.

Reply:
Oops, sorry, teach me not to read something while trying to listen to a meeting, even if I originally wrote it. :-) I'd thought we'd actually defined ufloat16 somewhere, but obviously not. I would certainly have expected an 11-bit mantissa, which would misalign things relative to the signed version and probably cause trouble.

That said (and sorry, late on a Friday...) float16 is only relevant if someone wants to think about what to encode in a void-extent block - otherwise "it's a float" is all we need to know, and 10.2 and 10.3 do set the precedent for unsigned floating point numbers. I now worry that I've multi-tasked enough that I've completely lost the thread. The translation via an actual float16 is an implementation detail for the decoding process, not a description of the format, no?

And you are, of course, right about the mental model for sampleLower. Sorry, it's been a while.

Reply:
> otherwise "it's a float" is all we need to know

No. ASTC HDR decoder's output is specifically `float16` (unless `rgb9e5` decoding mode is enabled). For example, this means that an ASTC HDR texture block cannot yield any value that is not exactly representable with the `float16` type. That said, DFD for UASTC HDR does not say anything about exact mantissa/exponent configuration of the decoded values so it's the color model's responsibility to provide appropriate normative definitions.

---

The question of whether to set the signed qualifier bit or not should depend on how UASTC HDR spec would handle void-extent blocks with negative values. It's easy to say that they're "forbidden" or "undefined" but many other block-compressed formats (including UASTC LDR and original ASTC) have well-defined fallback behavior for invalid inputs.

Since transcoding UASTC HDR to BC6H cannot preserve negative values, it would be reasonable to ensure that transcoding an arbitrary (including invalid) UASTC HDR payload to ASTC or 16-bit floats will also never yield negative values.

With such a guarantee in place, we'll be able to set the `SIGNED` qualifier to false, set `sampleLower` to `0.0`, and add required normative language to the color model definition.

Reply:
The [UASTC HDR spec](https://github.com/BinomialLLC/basis_universal/wiki/UASTC-HDR-Texture-Specification-v1.0) says the following things:

1. Positive inputs and outputs only. (This is not a limitation of UASTC HDR, but ASTC HDR. BC6H supports a signed variant, but ASTC HDR doesn't.) Transcoding is only supported to DXGI_FORMAT_BC6H_UF16.
2. _ASTC HDR Features Supported_ includes "Lossless solid color blocks (ASTC's HDR “void-extent” blocks). LDR void-extent blocks are not supported, only HDR (i.e. the ASTC header's "Dynamic Range" flag must be set to 1, for FP16 values)."
3. _Forbidden ASTC HDR Features_ does not list void-extent blocks with negative values.
4. Void-extent (solid color) blocks are immediately transcoded to solid color BC6H blocks using BC6H mode 14.
5. Void-Extent blocks must always use the HDR FP16 encoding (i.e. Bit 9, the Dynamic Range flag, must always be 1). The FP16 values cannot be negative, NaN or Inf.

I presume the decision reflected in #1 was made because negative solid color blocks is not enough functionality to justify also supporting transcoding to the BC6H signed variant.

I did not see any description of the handling if negative, NaN or Inf values are encountered. @richgel999 how does the transcoder handle this?

> Since transcoding UASTC HDR to BC6H cannot preserve negative values, it would be reasonable to ensure that transcoding an arbitrary (including invalid) UASTC HDR payload to ASTC or 16-bit floats will also never yield negative values.

AIUI no transcoding step is done from UASTC HDR to ASTC HDR. Therefore it is possible that a UASTC HDR image with an erroneous negative void-extent block could be decoded as ASTC producing a negative FP value.

Reply:
1. Semantically, the transcoder is just an implementation. The KDFS edition of the spec must define output for invalid data.
2. As of now, only the BC6H transcoder enforces all UASTC HDR restrictions and exits early when encountering an invalid block (including void-extent blocks with negative, NaN, or Inf values). Weaker validation of transcoding to other target formats is a bug, because it leads to non-portable behavior.

Reply:
Given that the decoded value has a sign bit, even though it is always 0 for valid UASTC HDR, the DFD sample's SIGNED attribute should be set. That leaves `sampleLower` to be decided.

Reply:
>That leaves `sampleLower` to be decided.

It **must** be `-1.0` if the `SIGNED` qualifier is set, otherwise decoded values would be incorrect.

Reply:
> > otherwise "it's a float" is all we need to know
> 
> No. ASTC HDR decoder's output is specifically `float16` (unless `rgb9e5` decoding mode is enabled). For example, this means that an ASTC HDR texture block cannot yield any value that is not exactly representable with the `float16` type. That said, DFD for UASTC HDR does not say anything about exact mantissa/exponent configuration of the decoded values so it's the color model's responsibility to provide appropriate normative definitions.

ASTC HDR mode is defined in terms of generating a (signed) float16 value, and since the implementation is required to be bit exact, there will have been a transcoding via that format, but the representation typically used in shaders is orthogonal - so you could argue it's no more part of the the decode process than any other internal representation, the description was just defined to plug into people's existing float16 hardware pipelines. As far as the DFD is concerned, I wouldn't lose sleep over the mention of float16 in the ASTC definition - it's just convention, and an indicator that the representation can represent values outside the 0..1 range and has some nonlinearity. BC6h is similar.

> The question of whether to set the signed qualifier bit or not should depend on how UASTC HDR spec would handle void-extent blocks with negative values. It's easy to say that they're "forbidden" or "undefined" but many other block-compressed formats (including UASTC LDR and original ASTC) have well-defined fallback behavior for invalid inputs.
> 
> Since transcoding UASTC HDR to BC6H cannot preserve negative values, it would be reasonable to ensure that transcoding an arbitrary (including invalid) UASTC HDR payload to ASTC or 16-bit floats will also never yield negative values.
> 
> With such a guarantee in place, we'll be able to set the `SIGNED` qualifier to false, set `sampleLower` to `0.0`, and add required normative language to the color model definition.

I don't know whether there would be a measurable performance impact in an implementation checking for negative void extent blocks - but it would be nice for users to know whether they manually need to filter negative values from textures, whether they might reliably make it to an ASTC target if that's all they care about, and whether we want to enforce this behaviour from all implementations. I could believe users who just want supercompressed ASTC might actually be happy to preserve negative void extent blocks (although they're very much the minority case). Obviously allowing people to use textures which *might* cause the BC6H decoder to break (as opposed to falling back to an error colour) is queueing up trouble if people start sharing UASTC content, and arguably it's a big ask for content creation applications to have to do their own ASTC checking to see whether the input is valid - if the encoder can be guaranteed to do this check and simply refuse to encode negative values, that may be the better solution, and probably be the less performance-critical part of the pipeline to start worrying about it.

TL;DR: If the BC6H decoder actually falls over when fed a negative void-extent block, that *is* a restriction on UASTC, not ASTC - although obviously the ASTC implementation of negative values is less flexible/useful than BC6H's for negatives, so I'm not sure it's worth "enhancing" the UASTC decoder to support signed BC6H outputs. The limitations of one output format doesn't mandate everything that could be represented (otherwise we'd be stuck with ETC1S's range), but not having a graceful fallback does.

Unless there's a plan to expand the range of the "standard" UASTC transcoders, I'm leaning towards describing UASTC here as unsigned float, with a 0..1.0 range, and a note requiring that ASTC void extent blocks be positive (hopefully checked in the encoder) - and perhaps that this does not affect the way floating point values are represented in ASTC void extent blocks for transcoders.

Reply:
> it's just convention, and an indicator that the representation can represent values outside the 0..1 range and has some nonlinearity

It's certainly more than that. Specifically, bit-exact ASTC decoding implies that all values that could be produced by the decoder (ignoring linear filtering) belong to the predefined set of 31744 finite non-negative `float16` values.

> I could believe users who just want supercompressed ASTC might actually be happy to preserve negative void extent blocks

Then they should use regular ASTC with Zstd and RDO.

> if the encoder can be guaranteed to do this check and simply refuse to encode negative values

The encoder does reject negative inputs but that's irrelevant here because the spec defines the decoding process.

> If the BC6H decoder actually falls over when fed a negative void-extent block

The BC6H transcoder immediately exits on encountering an ASTC block that it does not like. This includes void-extent blocks with negative or non-finite values as well as all other unsupported ASTC configurations (certain partitions, endpoint modes, etc).

> The limitations of one output format doesn't mandate everything that could be represented

In this case, they do mandate because UASTC HDR was created for a single transcoding target - BC6H unsigned. Despite name similarity, it has nothing to do with other Basis Universal codecs.

---

All that said, I too think that DFD should not set the signed bit and therefore the `sampleLower` must be 0.0.

Reply:
> > it's just convention, and an indicator that the representation can represent values outside the 0..1 range and has some nonlinearity
> 
> It's certainly more than that. Specifically, bit-exact ASTC decoding implies that all values that could be produced by the decoder (ignoring linear filtering) belong to the predefined set of 31744 finite non-negative `float16` values.

Oh, I just meant that the "float" in the DFD meant that. The actual definition of ASTC does indeed make more specific statements about the result, but the only way the DFD indicates that is by the colour model - there's nothing that says "16-bit float" that you could determine from the DFD without understanding that. Since you need to know the colour model to make sense of it anyway, the sign and float bit are only hints to a decoder about what range the results may fall in - and they don't say how far outside the 0..1 range things fall.

> > I could believe users who just want supercompressed ASTC might actually be happy to preserve negative void extent blocks
> 
> Then they should use regular ASTC with Zstd and RDO.

I was entirely being devil's advocate there. Sure, I'm happy to say that any user in that category has another way to achieve it.

> > if the encoder can be guaranteed to do this check and simply refuse to encode negative values
> 
> The encoder does reject negative inputs but that's irrelevant here because the spec defines the decoding process.

Okay, good. So if negative values aren't accepted by the (only conformant at this point?) encoder we can't reasonably say the format supports negative values.

> > If the BC6H decoder actually falls over when fed a negative void-extent block
> 
> The BC6H transcoder immediately exits on encountering an ASTC block that it does not like. This includes void-extent blocks with negative or non-finite values as well as all other unsupported ASTC configurations (certain partitions, endpoint modes, etc).

I think I stand by "it doesn't support negative values, and that's orthogonal to its opinions on whether ASTC does", then. :-)

> > The limitations of one output format doesn't mandate everything that could be represented
> 
> In this case, they do mandate because UASTC HDR was created for a single transcoding target - BC6H unsigned. Despite name similarity, it has nothing to do with other Basis Universal codecs.

Ah, a nuance which was lost on me - sorry, I figured it would still support other transcoding targets, although I suppose there was no reason to assume that and they would often not be useful.

> All that said, I too think that DFD should not set the signed bit and therefore the `sampleLower` must be 0.0.

I think that's consensus (unless Rich has strong feelings).

Reply:
>>    All that said, I too think that DFD should not set the signed bit and therefore the sampleLower must be 0.0.
>
> I think that's consensus (unless Rich has strong feelings).

Clearly I am (still) not understanding what a DFD sample is supposed to show for block-compressed color models. This is partly due to lack of explanation in KDFS that I previously discussed in private e-mail with @fluppeteer. In the none block-compressed cases it shows how to interpret the bits found in memory so if there is a sign bit, the SIGNED attribute must be set. The values used in the specified numerical range are irrelevant. I thought in the block-compressed case it was supposed to show what is the expected output from the decoder. In the case of using an ASTC decoder on the UASTC HDR data that is a float16 with a sign bit. Note  that the UASTC HDR transcoder is not a decoder.

Am I mistaken about the DFD sample intent for block-compressed textures?

Who/What are we trying to help with the sample information in this case?

Reply:
The definitions of sample information fields are already overridden and inconsistent for many current compressed color models:
- For BC1, the channel id is used to distinguish between alpha and opaque variants of the same sample with an extra restriction that only one channel can be present. One could argue that these should be two separate color models.
- For BC4/BC5 and R11/RG11, the signed qualifier is technically used to control the decoder, not to interpret its output.
- Same for BC6H, the signed qualifier changes the decoder behavior while the output is always `float16` that includes the sign bit.

Reply:
Thanks @lexaknyazev. I want to hear from @fluppeteer his intent for the sample descriptors for block-compressed color models before I make a decision. I don't think any changes are required to this PR as it refers to KDFS for the sample description, except for saying the FLOAT qualifier must be set. I do however need to notify @richgel999 so he can update his implementation.

Reply:
Response coming shortly as soon as I escape meetings (just so you know I'm responding). Sorry, cat with cystitis and epic timing.

Reply:
The DFD is intended to describe the representation and, therefore, range of the values which can be encoded. The decoding process description for ASTC HDR includes a signed float16 representation, but if it can't represent negative values, my feeling is it shouldn't be setting a signed bit. (@MarkCallow - could you remind me of the date of the email thread please? I'll try to clarify.)

BC1 was partly convenience that punch-through and default are almost the same format, and polluting the model enumerant space seemed unnecessary; it's slightly a shame to have to devote 8 bits to it anyway.

The signed qualifier on BC4/5 and R11/RG11 indicates whether the encoded values are in the 0..1 or -1..1 range - there is no "format of the output of the decoder". I think that's okay?

Likewise, BC6H (like ASTC HDR) has a decoding process that generates a float16 bit pattern, which an implementation then interprets to get the required number. "The resulting bit pattern should be interpreted as a 16-bit half float" doesn't mean "you must put the result in a 16-bit half float variable", it means "we've generated this bit pattern, and to get the number you want, that's what you need to interpret according to the float16 rules". The qualifier determines what values can be encoded in the format - and since that change (if only for efficiency) changes the process in the decoder, it also affects the described decode process - but the description in the KDFS is entirely "this is how you should interpret the content of the image", not "here is how some reference implementation behaves".

Reply:
My point is that there is a subtle semantic difference in how DFD is used for compressed and uncompressed formats.

In case of uncompressed formats, DFD describes how to interpret existing pixel data and it's fair to say that it does not imply any type conversions, i.e., DFD for regular `RGBA16F` describes both the bit layout on disk and the effective pixel values.

But in case of compressed formats, DFD (particularly the signed qualifier) is often used to _configure_ the decoder before feeding it with compressed data. Therefore DFD indeed describes only the encoded data, not the decoder's output.

Following this logic, the signed qualifier for UASTC HDR must be unset because UASTC HDR cannot encode negative values (by definition). This still requires us to define _some_ behavior for invalid UASTC HDR payloads but such clarifications won't affect DFD.

Reply:
I think I disagree with your belief that there's a difference, but since we agree on leaving signed unset, I think I'm happy to agree to disagree. :-)

Reply:
I think this spec quote:
> this representation indicates the output of compressed **decoding**

has a slightly different meaning than this @fluppeteer's statement above:
> The DFD is intended to describe the representation and, therefore, range of the values which can be **encoded**.

In any case, we have a consensus on the `SIGNED` qualifier bit and and the `sampleLower` value for UASTC HDR so this thread seems resolved.

Reply:
> has a slightly different meaning than this @fluppeteer's statement above

Another case for a clarification to KDFS.

> this thread seems resolved.

I need to confirm Rich that his implementation is still doing this which is what I told him to do back on Sept. 11th.